internal/auth/oidc/repository_auth_method.go

```diff
@@ -2,9 +2,218 @@ package oidc
+	}
+	if IdTokenClaims == nil {
+		return nil, errors.New(errors.InvalidParameter, op, "missing ID Token claims")
+	}
+	if AccessTokenClaims == nil {
+		return nil, errors.New(errors.InvalidParameter, op, "missing Access Token claims")
+	}
+	pubId, err := newAccountId()
+	if err != nil {
+		return nil, errors.Wrap(err, op)
+	}
+	var iss, sub string
+	var ok bool
+	if iss, ok = IdTokenClaims["iss"].(string); !ok {
+		return nil, errors.New(errors.Unknown, op, "issuer is not present in ID Token, which should not be possible")
+	}
+	if sub, ok = IdTokenClaims["sub"].(string); !ok {
+		return nil, errors.New(errors.Unknown, op, "subject is not present in Access Token, which should not be possible")
+	}
+
+	columns := []string{"public_id", "auth_method_id", "issuer_id", "subject_id"}
+	values := []interface{}{pubId, am.PublicId, iss, sub}
+	var conflictClauses, fieldMasks, nullMasks []string
+
+	var foundEmail, foundName interface{}
+	switch {
+	case AccessTokenClaims["name"] != nil:
+		foundName = AccessTokenClaims["name"]
+		columns, values = append(columns, "full_name"), append(values, foundName)
+	case IdTokenClaims["name"] != nil:
+		foundName = IdTokenClaims["name"]
+		columns, values = append(columns, "full_name"), append(values, foundName)
+	default:
+		conflictClauses = append(conflictClauses, "full_name = NULL")
+		nullMasks = append(nullMasks, "Name")
+	}
+	switch {
+	case AccessTokenClaims["email"] != nil:
+		foundEmail = AccessTokenClaims["email"]
+		columns, values = append(columns, "email"), append(values, foundEmail)
+	case IdTokenClaims["email"] != nil:
+		foundEmail = IdTokenClaims["email"]
+		columns, values = append(columns, "email"), append(values, foundEmail)
+	default:
+		conflictClauses = append(conflictClauses, "email = NULL")
+		nullMasks = append(nullMasks, "Email")
+	}
+
+	if foundName != nil {
+		values = append(values, foundName)
+		conflictClauses = append(conflictClauses, fmt.Sprintf("full_name = $%d", len(values)))
+		fieldMasks = append(fieldMasks, "Name")
+	}
+	if foundEmail != nil {
+		values = append(values, foundEmail)
+		conflictClauses = append(conflictClauses, fmt.Sprintf("email = $%d", len(values)))
+		fieldMasks = append(fieldMasks, "Email")
+	}
+
+	issAsUrl, err := url.Parse(iss)
+	if err != nil {
+		return nil, errors.New(errors.Unknown, op, "unable to parse issuer", errors.WithWrap(err))
+	}
+	acctForOplog, err := NewAccount(am.PublicId, issAsUrl, sub)
+	if err != nil {
+		return nil, errors.Wrap(err, op, errors.WithMsg("unable to create new acct for oplog"))
+	}
+
+	if foundName != nil {
+		acctForOplog.FullName = foundName.(string)
+	}
+	if foundEmail != nil {
+		acctForOplog.Email = foundEmail.(string)
+	}
+
+	placeHolders := make([]string, 0, len(columns))
+	for colNum := range columns {
+		placeHolders = append(placeHolders, fmt.Sprintf("$%d", colNum+1))
+	}
+	query := fmt.Sprintf(acctUpsertQuery, strings.Join(columns, ", "), strings.Join(placeHolders, ", "), strings.Join(conflictClauses, ", "))
```

Comment:
Nicely done! This is a great example of the extra work it sometimes takes to prevent a SQL injection attack. I have to be honest, when I first saw `acctUpsertQuery`, I was worried you were building the query with regular string replacement but after I read through it I see you are building up a parameterized query. Again, nice job.